### PR TITLE
stabilise ?, attributes on stmts, deprecate Reflect

### PR DESCRIPTION
**Pull request by nrc**
r? @nikomatsakis

### CODE REVIEW

File: src/librustc/diagnostics.rs [hunk 1]
```diff
@@ -1327,30 +1327,6 @@ let x: i32 = "I am not a number!";
...
-
-```text
-<std macros>:5:8: 6:42 error: mismatched types:
- expected `()`,
-     found `core::result::Result<_, _>`
- (expected (),
-     found enum `core::result::Result`) [E0308]
-```
-
-`try!` returns a `Result<T, E>`, and so the function must. But `main()` has
```
Comment (Manishearth) [inline src/librustc/diagnostics.rs]: doesn't this text apply to `?` as well?

**Review by Manishearth — APPROVED** [review]
lgtm, not sure if libs/lang team folks need to sign off. r=me otherwise

File: src/libcore/macros.rs [hunk 2]
```diff
@@ -255,6 +255,9 @@ macro_rules! debug_assert_ne {
 /// Helper macro for reducing boilerplate code for matching `Result` together
 /// with converting downstream errors.
 ///
+/// Prefer using `?` syntax to `try!`. `?` is built in to the language and is
+/// more succinct than `try!`, it is the standard method for error propogation.
```
Comment (petrochenkov) [inline src/libcore/macros.rs]: propogation -> propagation

**Review by msjyoo — APPROVED** [review]

File: src/libcore/macros.rs (hunk 2 above)
Comment (frewsxcv) [inline src/libcore/macros.rs]: > `?` is built in to the language and is more succinct than `try!`, it is the standard method for error propogation.

These are two separate sentences and probably shouldn't be separated by a comma.

**Comment by frewsxcv** [comment]
For the sake of cross-linking, here's the link to the tracking issue: https://github.com/rust-lang/rust/issues/31436

File: src/librustc/diagnostics.rs (hunk 1 above)
Comment (nrc) [inline src/librustc/diagnostics.rs]: No, annoyingly you get an error message about the Carrier trait

**Comment by nrc** [comment]
Also cc #15701, #27749

**Comment by nikomatsakis** [comment]
This looks good to me, though I do think we ought to try and address #36988.

**Comment by bluss** [comment]
What's the type inference situation? It used to be that replacing `try!(x)` with `x?` would not infer as often.

Has this been cratered to look for such issues?

**Comment by nrc** [comment]
@bluss good point, I had mis-remembered this issue. We have indeed done a bunch of Crater runs, but to test whether `?` is forwards compatible with the `Carrier` trait, which requires some backwards incompatibility between `try!` and `?`. I don't think we can ever replace the implementation of `try!` in a backwards compatible way and that shouldn't be in this PR.

@bors: r-

**Comment by nikomatsakis** [comment]
@nrc 

> I don't think we can ever replace the implementation of try! in a backwards compatible way and that shouldn't be in this PR.

argh, I missed that this change was even in the PR! good catch @bluss

**Pull request merged**